### PR TITLE
Task to fill up the new annotation pk and user_id values

### PR DESCRIPTION
**Pull request by marcospri**
Fill up the new annotation, the new pk with a value from the new sequence and, given that we are going to go over all rows anyway, take the opportunity to fill up the new `user_id` FK.

### Testing

- Make sure you are in the last version of the DB `tox -e dev --run-command 'alembic upgrade head'`

- Check the local annotations:

`docker compose exec postgres psql -U postgres -c "select id, pk, user_id from annotation order by updated asc limit 10"`

```                 id                  | pk | user_id 
--------------------------------------+----+---------
 f057ca22-4afe-11ee-9762-bbbdbc59f9ee |    |        
 b43188b0-48da-11ee-b8c7-1384f80c2902 |    |        
 9cd1f10a-48da-11ee-914c-93702ada8a92 |    |        
 75022b22-48c6-11ee-8eb5-1f0ef005b932 |    |        
 c71d19b6-45b6-11ee-9bf6-d7b735fdc34d |    |        
 35b3f2a4-45b3-11ee-9bf6-fbb4c8f399b4 |    |        
 90747dfe-45b2-11ee-9bf6-8f825f8c34c8 |    |        
 76b9ffb0-45b2-11ee-9bf6-cfc13962be76 |    |        
 f5d52cba-45ae-11ee-9bf6-5703e9a86c98 |    |        
 5baaa666-45ad-11ee-9bf6-53d59524d771 |    |        
(10 rows)
```

- In `make shell`, run the new task

```
from h.tasks.annotations import fill_pk_and_user_id
fill_pk_and_user_id.apply_async()
```

wait for the task to finish on `make dev` output:

```
[2023-09-04 16:56:59,221: INFO/ForkPoolWorker-16] h.tasks.annotations.fill_pk_and_user_id[ac1eeb3d-57bd-4222-bc2d-a8ae47421a7f] Task h.tasks.annotations.fill_pk_and_user_id[ac1eeb3d-57bd-4222-bc2d-a8ae47421a7f] succeeded in 12.00388038699748s: None
```

- Check the annotations again, the should have values in the new columns now:

```
docker compose exec postgres psql -U postgres -c "select id, pk, user_id from annotation order by updated asc limit 10" 
                  id                  | pk | user_id 
--------------------------------------+----+---------
 d56f6f26-2c59-11ee-ba2c-fff7b98eb96d |  1 |      53
 f300996a-2c5f-11ee-9c33-0242ac130002 |  5 |      53
 f3002656-2c5f-11ee-9c33-0242ac130002 |  4 |      53
 f30098b6-2c5f-11ee-9c33-0242ac130002 |  3 |      53
 f30099c4-2c5f-11ee-9c33-0242ac130002 |  2 |      53
 fc46cb2a-2c5f-11ee-9c33-0242ac130002 | 25 |      53
 fc45923c-2c5f-11ee-9c33-0242ac130002 | 24 |      53
 fc4595ac-2c5f-11ee-9c33-0242ac130002 | 23 |      53
 fc4593d6-2c5f-11ee-9c33-0242ac130002 | 22 |      53
 fc46c148-2c5f-11ee-9c33-0242ac130002 | 21 |      53
(10 rows)
```

### CODE REVIEW

File: h/tasks/annotations.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            == func.split_part(func.split_part(Annotation.userid, "@", 1), ":", 2),
+        )
+        .where(Annotation.pk.is_(None))
+        .order_by(Annotation.updated.asc())
+        .limit(batch_size)
+    ).cte("annotations")
+
+    db.execute(
+        update(Annotation)
+        .values(pk=Annotation.pk_sequence.next_value(), user_id=annotations.c.user_id)
```
Comment (marcospri) [inline h/tasks/annotations.py:34]: Update the pk value form the sequence

Comment (marcospri) [inline h/tasks/annotations.py:34]: And user_id from the value from the join

File: h/db/__init__.py [hunk 2]
```diff
@@ -63,6 +63,8 @@ def init(engine, base=Base, should_create=False, should_drop=False, authority=No
         # extension.
         engine.execute('CREATE EXTENSION IF NOT EXISTS "uuid-ossp";')
         base.metadata.create_all(engine)
+        # Create sequences not used explicitly in a column here
+        engine.execute("CREATE SEQUENCE IF NOT EXISTS annotation_id_seq;")
```
Comment (seanh) [inline h/db/__init__.py:67]: The sequence can't be added to the `Annotation` model somehow?

File: h/tasks/annotations.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        db.query(Annotation.id.label("annotation_id"), User.id.label("user_id"))
+        .join(
+            User,
+            User.username
+            == func.split_part(func.split_part(Annotation.userid, "@", 1), ":", 2),
+        )
+        .where(Annotation.pk.is_(None))
+        .order_by(Annotation.updated.asc())
+        .limit(batch_size)
+    ).cte("annotations")
```
Comment (seanh) [inline h/tasks/annotations.py:30]: I think you might want to consider [with_for_update(skip_locked=True)](https://docs.sqlalchemy.org/en/14/orm/query.html#sqlalchemy.orm.Query.with_for_update) here. It does an SQL `SELECT FOR UPDATE SKIP LOCKED` query. It'll ensure that if there are two instances of this task running at the same time they don't select the same rows from the table.

This is what the job queue uses: https://github.com/hypothesis/h/blob/41597de80b8e30d7762ee92348d96c6017eea844/h/services/search_index/_queue.py#L229-L236

File: h/tasks/annotations.py (hunk 1 above)
Comment (seanh) [reply h/tasks/annotations.py:34]: I'm wondering what `pk_sequence.next_value()` will do if there are two instances of this task running at the same time? Might just work?

File: h/tasks/annotations.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+from sqlalchemy import func, update
+
+from h.celery import celery, get_task_logger
+from h.models import Annotation, User
+
+log = get_task_logger(__name__)
+
+
+@celery.task
```
Comment (seanh) [inline h/tasks/annotations.py:9]: Keep in mind that h's global Celery config has a [hard time limit of `240` secs](https://github.com/hypothesis/h/blob/07b22e03b906e50c607086163799cee11037ad5f/h/celery.py#L75) (4 mins) for task runs. If you're going to use h-periodic to run this task periodically and you only want one instance of the task to be running at a time then I think h-periodic must not run the task more than once every 4 mins.

Alternatively if this task is expected to complete much faster than 4 mins then you can add a shorter `soft_time_limit` and `time_limit` here in the `@celery.task` decorator and then h-periodic will be free to run the task more frequently.

**Review by seanh — APPROVED** [review]
Note that New Relic monitors Celery tasks as "transactions" so you'll be able to see the performance of this task there

File: h/db/__init__.py (hunk 2 above)
Comment (marcospri) [reply h/db/__init__.py:67]: I reckon once we start using it (eg as the server_default) this won't be necessary.

File: h/tasks/annotations.py (hunk 1 above)
Comment (marcospri) [reply h/tasks/annotations.py:35]: Yeah postgres does the right thing for sequences, they always increase, so next_value will always have a bigger number, they are not guaranteed to be consecutive of course.

File: h/tasks/annotations.py (hunk 3 above)
Comment (marcospri) [reply h/tasks/annotations.py]: Done :+1:

File: h/tasks/annotations.py (hunk 4 above)
Comment (marcospri) [reply h/tasks/annotations.py:9]: To be honest I don't know what to expect 30s? 1m? 5m?

will adjust after the first few runs.

**Pull request merged**